### ISSUE TITLE
Max Attention Logit Calculation Diff Transformer

### ISSUE THREAD

**Issue by faresobeid** [microsoft/unilm#1666]
Hi just wanted to start by saying amazing work on differential transformer!

Was wondering for Table 5 whether the fact that the max attention logit is pretty stable could mean that qk norm and/or tanh soft clamping is not needed, saving some overhead. Although then I wondered how you calculated this pre softmax attention logit, as in was it the maximum attention logit out of q1k1^T and q2k2^T or in some other way?

Thanks

**Comment by YTianZHU** [comment]
@faresobeid Hi, thanks for the questions. As outliers in attention logits is greatly reduced, we believe it's true that qk norm and tanh clamping are not needed. It can be verified with some experiments on models with such design.

The pre-softmax attention logit is collected using abs of all q1k1^T / sqrt(d) and q2k2^T / sqrt(d).

**Comment by faresobeid** [comment]
Thanks for answering that sounds amazing! If you don't mind me asking another question, how were the hidden states norms calculated (before or after adding on the residual)? Also, should the lr's for the lambda params change with scale (muP style) even though they are of constant size, or change because of the parametrisation?

**Comment by YTianZHU** [comment]
@faresobeid The hidden states outliers are collected at the output of each layer (a layer contains an attention and an FFN, each with their own residuals). So, to your question, it is collected after adding the residual of FFN.

As shown in the code, for the initialization of lambda params, we use std=0.1. The strategy follows usual practice of initialization (let std of values be 1), 0.1**2 * 128 =1.28 (take head_dim=128 as an approximation). We find the strategy can be applied to different model sizes and configurations.

**Comment by faresobeid** [comment]
Thanks for the reply! For the lambda lr I meant if that should change with scale even if that std init is constant across model sizes. Also I was wondering why in none of your implementations you have one call for an attention kernel but with v duplicated across heads, then split after rather than 2 calls for attention. Thanks again

**Comment by YTianZHU** [comment]
I see. Upon our experiments and observation, there is no need for it to change with scale. However, we haven't conducted specific studies into it.

For the second question, the way you say is also a feasible implementation way. It can obtain the same output with current implementation.

**Comment by faresobeid** [comment]
Ok makes sense thank you for answering!
